modules/javafx.graphics/src/main/java/javafx/concurrent/Task.java

```diff
@@ -1014,12 +1016,31 @@ protected void failed() { }
             // state flag will not be readable immediately after this call. However,
             // that would be the case anyway since these properties are not thread-safe.
             if (isFxApplicationThread()) {
+                switch (getState()) {
+                case FAILED:
+                case SUCCEEDED:
+                    // a finished or failed task retains its state
+                    return false;
+                }
+
                 setState(Worker.State.CANCELLED);
             } else {
-                runLater(() -> setState(Worker.State.CANCELLED));
+                AtomicBoolean rv = new AtomicBoolean(flag);
+                runLater(() -> {
+                    switch (getState()) {
+                    case FAILED:
+                    case SUCCEEDED:
+                        // a finished or failed task retains its state
+                        rv.set(false);
+                        break;
+                    default:
+                        setState(Worker.State.CANCELLED);
+                        break;
+                    }
+                });
+                return rv.get();
```

Comment:
This is ineffective since you don't wait for the runLater to execute (and waiting could lead to deadlock, so I wouldn't recommend that). I don't think there is anything better than unconditionally returning `flag` when not running on the FX app thread. That's what the current code does (and is what your proposed fix will do most of the time).

Reply:
good point!

In this particular case, returning `true` from `cancel()` when state will not be changed is acceptable.